File: simbatch.py
#!/usr/bin/env python3

import sys
import subprocess
import getopt

def usage():
	print("SimBatch: A batch simulator for BondMachine designs")
	print("")
	print("SimBatch allows you to run batch simulations of BondMachine designs using a CSV input file and producing a CSV output file.")
	print("It expects the BondMachine design to be already compiled in the working directory called `bondmachine.json`.")
	print("")
	print("Usage: simbatch [options]")
	print("Options:")
	print("  -w, --working-dir DIR        set the working directory (default: working_dir)")
	print("  -i, --input-file FILE        set the input CSV file (default: simbatch_input.csv)")
	print("  -o, --output-file FILE       set the output CSV file (default: working_dir/simbatch_output.csv)")
	print("  -s, --simulation-steps N     number of simulation steps (default: 200)")
	print("  -m, --ml                     enable ML output formatting (probabilities + classification)")
	print("  -b, --benchcore              enable benchcore mode")
	print("  -H, --header                 include header row in output CSV")
	print("  -P, --prefix                 include data type prefix in output CSV")
	print("  -d, --data-type TYPE         data type for outputs (e.g. float32) (default: float32)")
	print("  --linear-data-range RANGE    pass a linear data range option to bondmachine/bmnumbers")
	print("  -v, --stop-on-valid-of N     stop on valid of output index N")
	print("  -h, --help                   show this help message and exit")
	print("")
	print("Example:")
	print("  simbatch.py -w working_dir -i input.csv -o out.csv -s 200")

working_dir=""
input_file=""
output_file=""
simulation_steps="200"
isml=False
benchCore=False
stopOnValidOf=-1
linear_data_range=""
data_type="float32"
prefix="0f"
header=False
omit_prefix=True

# Parse command line options
try:
	opts, args = getopt.getopt(sys.argv[1:], "w:i:o:s:mbd:v:hH", ["working-dir=","input-file=","output-file=", "simulation-steps=","ml", "benchcore", "linear-data-range=","data-type=", "stop-on-valid-of=", "help", "header"])
except  getopt.GetoptError:
	usage()
	sys.exit(2)


for o, a in opts:
	if o in ("-w", "--working-dir"):
		working_dir = a
	elif o in ("-i", "--input-file"):
		input_file = a
	elif o in ("-o", "--output-file"):
		output_file = a
	elif o in ("-s", "--simulation-steps"):
		simulation_steps = a
	elif o in ("--linear-data-range"):
		linear_data_range = "-linear-data-range "+a
	elif o in ("-d", "--data-type"):
		data_type = a
	elif o in ("-m", "--ml"):
		isml = True
	elif o in ("-b", "--benchcore"):
		benchCore = True
	elif o in ("-h", "--help"):
		usage()
		sys.exit(0)
	elif o in ("-v", "--stop-on-valid-of"):
		stopOnValidOf = int(a)
	elif o in ("-H", "--header"):
		header = True
	elif o in ("-P", "--prefix"):
		omit_prefix = False

if working_dir == "":
	working_dir="working_dir"
if input_file == "":
	input_file="simbatch_input.csv"
if output_file == "":
	output_file=working_dir+"/simbatch_output.csv"

last_step = str(int(simulation_steps) - 1)

# Load the BondMachine inputs
command="bondmachine -bondmachine-file "+working_dir+"/bondmachine.json -list-inputs "+ linear_data_range
p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, close_fds=True)
p.wait()
inputs={}
if p.returncode==0:
	while True:
		o = p.stdout.readline().decode()
		if o == '' and p.poll() != None:
			break
		result=o.split()
		if len(result)==2:
			inputs[result[0]]=result[1]

# Load the BondMachine outputs
command="bondmachine -bondmachine-file "+working_dir+"/bondmachine.json -list-outputs "+linear_data_range
p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, close_fds=True)
p.wait()
outputs={}
if p.returncode==0:
	while True:
		o = p.stdout.readline().decode()
		if o == '' and p.poll() != None:
			break
		result=o.split()
		if len(result)==2:
			outputs[result[0]]=result[1]

# Get the number prefix
command="bmnumbers -get-prefix " + data_type +" "+linear_data_range
p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, close_fds=True)
p.wait()
if p.returncode==0:
	while True:
		o = p.stdout.readline().decode()
		if o == '' and p.poll() != None:
			break
		prefix=o.strip()

# print (prefix)
# print (inputs)
# print (outputs)

# Open the input file
input_file_handle=open(input_file, "r")
output_file_handle=open(output_file, "w")

# Write the output file header if needed
if header:
	if isml:
		for i in range(len(outputs)):
			output_file_handle.write("probability_"+str(i)+",")
		output_file_handle.write("classification")
	if benchCore:
		output_file_handle.write(",latency_cycles")
	output_file_handle.write("\n")

# Read every line of the input file
for line in input_file_handle:
	line=line.strip()
	inputs_values=line.split(",")
	if len(inputs_values)==0:
		continue
	elif len(inputs_values)==len(inputs):
		print ("Running simulation with inputs: "+line)

		# Remove the simbox file
		command="rm -f "+working_dir+"/simboxtemp.json"
		p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, close_fds=True)
		p.wait()

		# Prepare the simbox general commands, like showing IOs, ticks, pc, disasm, etc.
		# These are useful for debugging and are disabled by default
		commands=[]
		commands.append("simbox -simbox-file "+working_dir+"/simboxtemp.json -add \"config:show_io_pre\"")
		commands.append("simbox -simbox-file "+working_dir+"/simboxtemp.json -suspend 0") 
		commands.append("simbox -simbox-file "+working_dir+"/simboxtemp.json -add \"config:show_io_post\"")
		commands.append("simbox -simbox-file "+working_dir+"/simboxtemp.json -suspend 1")
		commands.append("simbox -simbox-file "+working_dir+"/simboxtemp.json -add \"config:show_ticks\"")
		commands.append("simbox -simbox-file "+working_dir+"/simboxtemp.json -suspend 2")
		commands.append("simbox -simbox-file "+working_dir+"/simboxtemp.json -add \"config:show_pc\"")
		commands.append("simbox -simbox-file "+working_dir+"/simboxtemp.json -suspend 3")
		commands.append("simbox -simbox-file "+working_dir+"/simboxtemp.json -add \"config:show_disasm\"")
		commands.append("simbox -simbox-file "+working_dir+"/simboxtemp.json -suspend 4")

		for command in commands:
			p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, close_fds=True)
			p.wait()
			if p.returncode!=0:
				print ("Error preparing simbox command: "+command)
				sys.exit(2)

		# Add the inputs to the simbox
		for i in range(len(inputs_values)):
			input_name=inputs[str(i)]
			input_value=inputs_values[i]
			command="simbox -simbox-file "+working_dir+"/simboxtemp.json -add \"absolute:0:set:"+input_name+":"+input_value+"\""
			p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, close_fds=True)
			p.wait()
			if p.returncode!=0:
				print ("Error setting input "+input_name+" to "+input_value)
				sys.exit(2)

		# Prepare the outputs to be collected
		for output_name in outputs:
			if benchCore and outputs[output_name]=="o"+str(len(outputs)-1):
				command="simbox -simbox-file "+working_dir+"/simboxtemp.json -add \"onexit:show:"+outputs[output_name]+":unsigned\""
			else:
				command="simbox -simbox-file "+working_dir+"/simboxtemp.json -add \"onexit:show:"+outputs[output_name]+":"+data_type+"\""
			p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, close_fds=True)
			p.wait()
			if p.returncode!=0:
				print ("Error getting output "+output_name)
				sys.exit(2)

		stopOnValidOf=len(outputs)-1

		# Run the simulation
		command="bondmachine -bondmachine-file "+working_dir+"/bondmachine.json -simbox-file "+working_dir+"/simboxtemp.json -sim-stop-on-valid-of "+str(stopOnValidOf)+" -sim -sim-interactions "+simulation_steps+" "+linear_data_range
		p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, close_fds=True)
		p.wait()
		if p.returncode!=0:
			print ("Error running simulation")
			print (p.stderr.read().decode())
			sys.exit(2)

		outline=p.stdout.read().decode().strip()
		if omit_prefix:
			outline=outline.replace(prefix,"")

		# If there is an active benchCore mode, extract the latency cycles that is the last output
		if benchCore:
			parts=outline.split(' ')
			latency_cycles=parts[-1]
			outline=" ".join(parts[:-1])

		if isml:
			import numpy as np
			vals=np.asarray(outline.split(' '))
			vals=vals.astype(np.float32) # TODO: make this configurable
			index=np.argmax(vals)
			outline=outline.replace(" ",",")
			outline=outline+ "," + str(index)
		else:
			outline=outline.strip(',')
			outline=outline.replace(" ",",")

		# Write the output line eventually adding the latency cycles at the end
		output_file_handle.write(outline)
		if benchCore:
			output_file_handle.write(","+latency_cycles)
		output_file_handle.write("\n")

	else:
		print("Error: The input file has an invalid number of columns")

input_file_handle.close()
output_file_handle.close()

File: simbatch.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
)

type Config struct {
	workingDir       string
	inputFile        string
	outputFile       string
	simulationSteps  string
	isML             bool
	benchCore        bool
	stopOnValidOf    int
	linearDataRange  string
	dataType         string
	prefix           string
	header           bool
	omitPrefix       bool
	delaysFile       string
	delayString      string
}

func usage() {
	fmt.Println("SimBatch: A batch simulator for BondMachine designs")
	fmt.Println("")
	fmt.Println("SimBatch allows you to run batch simulations of BondMachine designs using a CSV input file and producing a CSV output file.")
	fmt.Println("It expects the BondMachine design to be already compiled in the working directory called `bondmachine.json`.")
	fmt.Println("")
	fmt.Println("Usage: simbatch [options]")
	fmt.Println("Options:")
	fmt.Println("  -w, --working-dir DIR        set the working directory (default: working_dir)")
	fmt.Println("  -i, --input-file FILE        set the input CSV file (default: simbatch_input.csv)")
	fmt.Println("  -o, --output-file FILE       set the output CSV file (default: working_dir/simbatch_output.csv)")
	fmt.Println("  -s, --simulation-steps N     number of simulation steps (default: 200)")
	fmt.Println("  -m, --ml                     enable ML output formatting (probabilities + classification)")
	fmt.Println("  -b, --benchcore              enable benchcore mode")
	fmt.Println("  -H, --header                 include header row in output CSV")
	fmt.Println("  -P, --prefix                 include data type prefix in output CSV")
	fmt.Println("  -d, --data-type TYPE         data type for outputs (e.g. float32) (default: float32)")
	fmt.Println("  -l, --linear-data-range RANGE    pass a linear data range option to bondmachine/bmnumbers")
	fmt.Println("  -v, --stop-on-valid-of N     stop on valid of output index N")
	fmt.Println("  -h, --help                   show this help message and exit")
	fmt.Println("  -y, --delays-file FILE       set the delays file")
	fmt.Println("")
	fmt.Println("Example:")
	fmt.Println("  simbatch -w working_dir -i input.csv -o out.csv -s 200")
}

func runCommand(command string) (string, error) {
	cmd := exec.Command("sh", "-c", command)
	output, err := cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("command failed: %v, output: %s", err, string(output))
	}
	return string(output), nil
}

func loadInputsOrOutputs(workingDir, option, linearDataRange string) (map[string]string, error) {
	command := fmt.Sprintf("bondmachine -bondmachine-file %s/bondmachine.json %s %s", workingDir, option, linearDataRange)
	output, err := runCommand(command)
	if err != nil {
		return nil, err
	}

	result := make(map[string]string)
	scanner := bufio.NewScanner(strings.NewReader(output))
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Fields(line)
		if len(parts) == 2 {
			result[parts[0]] = parts[1]
		}
	}
	return result, scanner.Err()
}

func getPrefix(dataType, linearDataRange string) (string, error) {
	command := fmt.Sprintf("bmnumbers -get-prefix %s %s", dataType, linearDataRange)
	output, err := runCommand(command)
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(output), nil
}

func findMaxIndex(values []float32) int {
	if len(values) == 0 {
		return -1
	}
	maxIdx := 0
	maxVal := values[0]
	for i, v := range values {
		if v > maxVal {
			maxVal = v
			maxIdx = i
		}
	}
	return maxIdx
}

func main() {
	cfg := Config{
		simulationSteps: "200",
		dataType:        "float32",
		prefix:          "0f",
		omitPrefix:      true,
		stopOnValidOf:   -1,
	}

	// Parse command-line flags
	flag.StringVar(&cfg.workingDir, "w", "", "working directory")
	flag.StringVar(&cfg.workingDir, "working-dir", "", "working directory")
	flag.StringVar(&cfg.inputFile, "i", "", "input CSV file")
	flag.StringVar(&cfg.inputFile, "input-file", "", "input CSV file")
	flag.StringVar(&cfg.outputFile, "o", "", "output CSV file")
	flag.StringVar(&cfg.outputFile, "output-file", "", "output CSV file")
	flag.StringVar(&cfg.simulationSteps, "s", "200", "simulation steps")
	flag.StringVar(&cfg.simulationSteps, "simulation-steps", "200", "simulation steps")
	flag.BoolVar(&cfg.isML, "m", false, "enable ML output")
	flag.BoolVar(&cfg.isML, "ml", false, "enable ML output")
	flag.BoolVar(&cfg.benchCore, "b", false, "enable benchcore")
	flag.BoolVar(&cfg.benchCore, "benchcore", false, "enable benchcore")
	flag.BoolVar(&cfg.header, "H", false, "include header")
	flag.BoolVar(&cfg.header, "header", false, "include header")
	var includePrefix bool
	flag.BoolVar(&includePrefix, "P", false, "include prefix")
	flag.BoolVar(&includePrefix, "prefix", false, "include prefix")
	flag.StringVar(&cfg.dataType, "d", "float32", "data type")
	flag.StringVar(&cfg.dataType, "data-type", "float32", "data type")
	var linearDataRangeArg string
	flag.StringVar(&linearDataRangeArg, "l", "", "linear data range")
	flag.StringVar(&linearDataRangeArg, "linear-data-range", "", "linear data range")
	flag.IntVar(&cfg.stopOnValidOf, "v", -1, "stop on valid of")
	flag.IntVar(&cfg.stopOnValidOf, "stop-on-valid-of", -1, "stop on valid of")
	flag.StringVar(&cfg.delaysFile, "y", "", "delays file")
	flag.StringVar(&cfg.delaysFile, "delays-file", "", "delays file")
	helpFlag := flag.Bool("h", false, "show help")
	helpFlagLong := flag.Bool("help", false, "show help")

	flag.Parse()

	if *helpFlag || *helpFlagLong {
		usage()
		os.Exit(0)
	}

	// Set defaults
	if cfg.workingDir == "" {
		cfg.workingDir = "working_dir"
	}
	if cfg.inputFile == "" {
		cfg.inputFile = "simbatch_input.csv"
	}
	if cfg.outputFile == "" {
		cfg.outputFile = cfg.workingDir + "/simbatch_output.csv"
	}
	if linearDataRangeArg != "" {
		cfg.linearDataRange = "-linear-data-range " + linearDataRangeArg
	}
	if cfg.delaysFile != "" {
		cfg.delayString = "-sim-delays-file " + cfg.delaysFile
	}
	if includePrefix {
		cfg.omitPrefix = false
	}

	// Load BondMachine inputs
	inputs, err := loadInputsOrOutputs(cfg.workingDir, "-list-inputs", cfg.linearDataRange)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error loading inputs: %v\n", err)
		os.Exit(1)
	}

	// Load BondMachine outputs
	outputs, err := loadInputsOrOutputs(cfg.workingDir, "-list-outputs", cfg.linearDataRange)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error loading outputs: %v\n", err)
		os.Exit(1)
	}

	// Get number prefix
	prefix, err := getPrefix(cfg.dataType, cfg.linearDataRange)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error getting prefix: %v\n", err)
		os.Exit(1)
	}
	cfg.prefix = prefix

	// Open input file
	inputFileHandle, err := os.Open(cfg.inputFile)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening input file: %v\n", err)
		os.Exit(1)
	}
	defer inputFileHandle.Close()

	// Create output file
	outputFileHandle, err := os.Create(cfg.outputFile)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error creating output file: %v\n", err)
		os.Exit(1)
	}
	defer outputFileHandle.Close()

	writer := bufio.NewWriter(outputFileHandle)
	defer writer.Flush()

	// Write header if needed
	if cfg.header {
		if cfg.isML {
			for i := 0; i < len(outputs); i++ {
				fmt.Fprintf(writer, "probability_%d,", i)
			}
			fmt.Fprint(writer, "classification")
		}
		if cfg.benchCore {
			fmt.Fprint(writer, ",latency_cycles")
		}
		fmt.Fprintln(writer)
	}

	// Process each line
	scanner := bufio.NewScanner(inputFileHandle)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}

		inputsValues := strings.Split(line, ",")
		if len(inputsValues) == 0 {
			continue
		}

		if len(inputsValues) != len(inputs) {
			fmt.Fprintf(os.Stderr, "Error: The input file has an invalid number of columns\n")
			continue
		}

		fmt.Printf("Running simulation with inputs: %s\n", line)

		// Remove simbox file
		simboxFile := cfg.workingDir + "/simboxtemp.json"
		os.Remove(simboxFile)

		// Prepare simbox general commands
		commands := []string{
			fmt.Sprintf("simbox -simbox-file %s -add \"config:show_io_pre\"", simboxFile),
			fmt.Sprintf("simbox -simbox-file %s -suspend 0", simboxFile),
			fmt.Sprintf("simbox -simbox-file %s -add \"config:show_io_post\"", simboxFile),
			fmt.Sprintf("simbox -simbox-file %s -suspend 1", simboxFile),
			fmt.Sprintf("simbox -simbox-file %s -add \"config:show_ticks\"", simboxFile),
			fmt.Sprintf("simbox -simbox-file %s -suspend 2", simboxFile),
			fmt.Sprintf("simbox -simbox-file %s -add \"config:show_pc\"", simboxFile),
			fmt.Sprintf("simbox -simbox-file %s -suspend 3", simboxFile),
			fmt.Sprintf("simbox -simbox-file %s -add \"config:show_disasm\"", simboxFile),
			fmt.Sprintf("simbox -simbox-file %s -suspend 4", simboxFile),
		}

		for _, cmd := range commands {
			if _, err := runCommand(cmd); err != nil {
				fmt.Fprintf(os.Stderr, "Error preparing simbox command: %s\n%v\n", cmd, err)
				os.Exit(2)
			}
		}

		// Add inputs to simbox
		for i := 0; i < len(inputsValues); i++ {
			inputName := inputs[strconv.Itoa(i)]
			inputValue := inputsValues[i]
			cmd := fmt.Sprintf("simbox -simbox-file %s -add \"absolute:0:set:%s:%s\"", simboxFile, inputName, inputValue)
			if _, err := runCommand(cmd); err != nil {
				fmt.Fprintf(os.Stderr, "Error setting input %s to %s\n%v\n", inputName, inputValue, err)
				os.Exit(2)
			}
		}

		// Prepare outputs
		for outputName, outputVal := range outputs {
			var cmd string
			if cfg.benchCore && outputVal == "o"+strconv.Itoa(len(outputs)-1) {
				cmd = fmt.Sprintf("simbox -simbox-file %s -add \"onexit:show:%s:unsigned\"", simboxFile, outputVal)
			} else {
				cmd = fmt.Sprintf("simbox -simbox-file %s -add \"onexit:show:%s:%s\"", simboxFile, outputVal, cfg.dataType)
			}
			if _, err := runCommand(cmd); err != nil {
				fmt.Fprintf(os.Stderr, "Error getting output %s\n%v\n", outputName, err)
				os.Exit(2)
			}
		}

		cfg.stopOnValidOf = len(outputs) - 1

		// Run simulation
		simCmd := fmt.Sprintf("bondmachine -bondmachine-file %s/bondmachine.json %s -simbox-file %s -sim-stop-on-valid-of %d -sim -sim-interactions %s %s",
			cfg.workingDir, cfg.delayString, simboxFile, cfg.stopOnValidOf, cfg.simulationSteps, cfg.linearDataRange)
		simOutput, err := runCommand(simCmd)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error running simulation\n%v\n", err)
			os.Exit(2)
		}

		outline := strings.TrimSpace(simOutput)
		if cfg.omitPrefix {
			outline = strings.ReplaceAll(outline, cfg.prefix, "")
		}

		// Handle benchCore mode
		var latencyCycles string
		if cfg.benchCore {
			parts := strings.Fields(outline)
			if len(parts) > 0 {
				latencyCycles = parts[len(parts)-1]
				outline = strings.Join(parts[:len(parts)-1], " ")
			}
		}

		// Handle ML mode
		if cfg.isML {
			vals := strings.Fields(outline)
			floatVals := make([]float32, len(vals))
			for i, v := range vals {
				f, _ := strconv.ParseFloat(v, 32)
				floatVals[i] = float32(f)
			}
			maxIdx := findMaxIndex(floatVals)
			outline = strings.ReplaceAll(outline, " ", ",")
			outline = fmt.Sprintf("%s,%d", outline, maxIdx)
		} else {
			outline = strings.Trim(outline, ",")
			outline = strings.ReplaceAll(outline, " ", ",")
		}

		// Write output
		fmt.Fprint(writer, outline)
		if cfg.benchCore {
			fmt.Fprintf(writer, ",%s", latencyCycles)
		}
		fmt.Fprintln(writer)
	}

	if err := scanner.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input file: %v\n", err)
		os.Exit(1)
	}
}
